addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/SameOriginMethodExecutionScanRule.java

```diff
@@ -0,0 +1,227 @@
+    private static final Pattern JSONP_PATTERN =
+            Pattern.compile("([\\w$]+?)\\(.*?\\)", Pattern.CASE_INSENSITIVE);
+    private static final ArrayList<String> CALLBACK_PARAMS =
+            new ArrayList<String>() {
+                {
+                    add("callback");
+                    add("target");
+                    add("cb");
+                    add("jsonp");
+                    add("cmd");
+                    add("readyFunction");
+                    add("jsoncallback");
+                }
+            };
+    private static final Map<String, String> ALERT_TAGS =
+            CommonAlertTag.toMap(
+                    CommonAlertTag.OWASP_2021_A04_INSECURE_DESIGN,
+                    CommonAlertTag.OWASP_2017_A03_DATA_EXPOSED);
+
+    private static final Logger LOGGER =
+            LogManager.getLogger(SameOriginMethodExecutionScanRule.class);
+
+    @Override
+    public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
+        HttpMessage newReq = msg.cloneRequest();
+        String respBody = msg.getResponseBody().toString();
+        TreeSet<HtmlParameter> allUrlParameters = getUrlParameters(newReq);
+        List<Element> scriptTags = source.getAllElements(HTMLElementName.SCRIPT);
+        String sfdHeader = msg.getRequestHeader().getHeader("Sec-Fetch-Dest");
+        String refererHeader = msg.getRequestHeader().getHeader("Referer");
+
+        // Method 1: Heuristics check on <script> tag.
+        // Test URL parameters
+        if (msg.getResponseHeader().isHtml()) {
+            for (HtmlParameter param : allUrlParameters) {
+                if (param.getValue().isEmpty()) {
+                    continue;
+                }
+
+                performHeuristicsCheck(
+                        param,
+                        scriptTags,
+                        msg.getRequestHeader().getURI().toString().contains(".swf"));
+            }
+        }
+
+        // Method 2: Check JSONP endpoint
+        if (!msg.getResponseHeader().isJson()
+                || !msg.getRequestHeader().getMethod().equals(HttpRequestHeader.GET)
+                || !(msg.getResponseHeader().getStatusCode() == HttpStatusCode.OK)
+                || !(sfdHeader.equals("script") || sfdHeader == null)
+                || respBody.startsWith("jquery")) {
+            return;
+        }
+
+        try {
+            JSONObject.fromObject(respBody);
+            return;
+        } catch (JSONException e) {
+            // If execution reaches here it means that the response body was not valid JSON,
+            // which is likely to be the case in JSONP responses.
+        }
+
+        for (HtmlParameter param : allUrlParameters) {
+            Matcher funcMatcher = JSONP_PATTERN.matcher(respBody);
+            while (funcMatcher.find()) {
+                String matchedFuncName = funcMatcher.group(1);
+                if (matchedFuncName.equals(param.getValue())) {
+                    LOGGER.info(
```

Comment:
This can probably be moved to debug vs info.

Reply:
Removed.